[PT Run] Service Plugin

davidegiacometti opened the pull request:
## Summary of the Pull Request

PT Run plugin for search, start, stop and restart Windows services.
This is just a draft.

![PT Run Service](https://user-images.githubusercontent.com/25966642/101506327-888f9480-3975-11eb-88e8-0f1113a73c26.gif)

![UAC](https://user-images.githubusercontent.com/25966642/99307539-ef2e0080-2856-11eb-83e4-9962a71aaebc.png)
**UAC on stop/start/restart**

## PR Checklist
* [x] Applies to #6071
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #6071
- [x] **Tests:** Added/updated and all pass
- [x] **Installer:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [x] **Binaries:** Any new files are added to WXS / YML
   - [ ] No new binaries
   - [x] [YML for signing](https://github.com/microsoft/PowerToys/blob/master/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [x] [WXS for installer](https://github.com/microsoft/PowerToys/blob/master/installer/PowerToysSetup/Product.wxs) for new binaries
## Info on Pull Request

TO-DO

* [x] Icon
* [x] Localization
* [x] Setup
* [ ] ~~Improve UAC application name~~ (further implementation)
* [ ] ~~Open the selected service in Services~~ (not sure this can be done)

## Validation Steps Performed

Manually test the plugin

alannt777 commented:
New icon for services?

/cc @niels9001

alannt777 commented:
There are some icon ideas in #6071

crutkas commented:
spelled word should be `occurred`

crutkas commented:
for loc, it is removing hardcoded strings

niels9001 commented:
![services light](https://user-images.githubusercontent.com/9866362/99716016-de27fe00-2aa7-11eb-9a9a-a03b1bfedf89.png)
![services dark](https://user-images.githubusercontent.com/9866362/99716018-dec09480-2aa7-11eb-84db-f873ed4429e2.png)

@davidegiacometti I have created a light/dark version of a potential plug-in icon. This uses the MDL2 version of the Windows Services app icon.

davidegiacometti commented:
Thank you @niels9001 
I will work on the plugin this weekend 😃

crutkas commented:
fixed conflict

htcfreek commented:
@davidegiacometti 
1. Can we get rid of the message box? I personally don't like it.

2. And can we get rid of calling the cmd?
[I think it would be better if the UAC shows something like "PT Run Service Plugin (Helper)" instead of "Command Processor". I think that "Command Processor" sounds a bit malicious/danger at this moment when I am managing a service from PT Run. - Hope you understand what I trying to describe.]

crutkas commented:
What message prompt?  The UAC dialog?

htcfreek commented:
> What message prompt?  The UAC dialog?

No. In the giv you can see that the command gets answered by a message box dialogue which say something like everything is done or operation failed.

yuyoyuppe commented:
please rebase on the latest master, so the CI can pass.

davidegiacometti commented:
@htcfreek 
1 neither do I but should I just start/stop/restart the service without give a feedback to the user?
2 I understand. I was trying to manage services using ServiveController At first but since PowerLauncher doesn't run elevated isn't possible.

niels9001 commented:
@davidegiacometti @htcfreek I agree with both. There needs to be feedback to indicate that it actually worked, but a win32 MessageBox feels a bit harring.

I think a Windows 10 toast notification would be the best solution: it tells the user about the action, while light-dismissable. Other services in Windows use that as well (e.g. USB drive or Your Phone).

PowerToys Settings uses these toast  notifications, might be re-usable?

htcfreek commented:
> @davidegiacometti @htcfreek I agree with both. There needs to be feedback to indicate that it actually worked, but a win32 MessageBox feels a bit harring.
> 
> I think a Windows 10 toast notification would be the best solution: it tells the user about the action, while light-dismissable. Other services in Windows use that as well (e.g. USB drive or Your Phone).
> 
> PowerToys Settings uses these toast  notifications, might be re-usable?

I like this idea.

htcfreek commented:
@davidegiacometti 
2. I understand your problem. But you don't have to call cmd.exe because you can call `%windir%\system32\net.exe` directly. And I would create a `PTRun.ServicePlugin.Helper.exe`, call it elevated and let this exe than call `net.exe`. Then the UAC should show something like `PTRun Service Plugin Helper`.

crutkas commented:
@davidegiacometti this needs to be rebased against master due to VS 16.8.  I can't easily test until then.

@niels9001 can we get the base of this working and then tweak later?

niels9001 commented:
@crutkas Sure, nothing blocking.. thanks again @davidegiacometti !

davidegiacometti commented:
@niels9001 I like the idea of the toast notification but I am not sure we can reuse Settings one.
@htcfreek I have set `net.exe` for start and stop. I think the external executable could be the definitive solution.

enricogior commented:
@yuyoyuppe 
can you provide some directions on how to use toast notifications from within a PT Run plug-in?

yuyoyuppe commented:
We have a set of helper functions for using toast notifications in C++. It's part of the `common` static library located in [`notifications.h`](https://github.com/microsoft/PowerToys/blob/master/src/common/notifications.h). There're 2 options for reusing it from C#:

- build notifications.cpp/h as part of our C++/CLI glue library called `interop`
- do an RPC call from the PT Run process to the runner with toast title/contents

[Here's an example of how we use the API from the bootstrapper app](https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/installer/PowerToysBootstrapper/bootstrapper/bootstrapper.cpp#L236). 
- we need to call `notifications::override_application_id` + `notifications::register_application_id`, since toast notifications require an icon for the executable to work. 
- Afterwards, just call `notifications::show_toast(L"msg", L"title");` 

There's also an example of how to use [Windows Community Toolkit](https://github.com/WindowsNotifications/desktop-toasts) from a desktop C# app, perhaps it could be easier then exposing our API.

enricogior commented:
> There's also an example of how to use Windows Community Toolkit from a desktop C# app, perhaps it could be easier then exposing our API.

Definitely, since I also don't like the idea to create too many dependencies between projects.
The notification implementation in C++ is customized for the needs of the installer/updater.
We are going to make changes there depending on the needs, and having to support interop as well would only make the code maintainability more complicated.

davidegiacometti commented:
@enricogior @yuyoyuppe Thank you!
I have been able to send notifications from the Launcher plugin following [doc](https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/send-local-toast-desktop) and [example](https://github.com/WindowsNotifications/desktop-toasts).
I think the best choise is to implement notification in `IPublicAPI` and provide an API to send notification to every plugin that needs this.
Since PowerLauncher doesn't have a shortcut in Start menu I have registered the `NotificationActivator` to the AUMID `Microsoft.PowerToysWin32` that is already used by the bootstrapper, is this fine?

yuyoyuppe commented:
@davidegiacometti good job! 

Regarding the shortcut, I'd rather not rely on the bootstrapper's ID, since [we're removing the shortcut when the installation finishes](https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/installer/PowerToysBootstrapper/bootstrapper/bootstrapper.cpp#L246-L248), and IIRC, it should work from a different executable at all. Perhaps the Windows Community Toolkit registers the shortcut under the hood? Could you please try using a different app id?

davidegiacometti commented:
> @davidegiacometti good job!
> 
> Regarding the shortcut, I'd rather not rely on the bootstrapper's ID, since [we're removing the shortcut when the installation finishes](https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/installer/PowerToysBootstrapper/bootstrapper/bootstrapper.cpp#L246-L248), and IIRC, it should work from a different executable at all. Perhaps the Windows Community Toolkit registers the shortcut under the hood? Could you please try using a different app id?

Actually it's the runner activator and not the bootstrapper one and the shortcut is created during the installation: https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/installer/PowerToysSetup/Product.wxs#L308

According to the [doc](https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/send-local-toast-desktop?tabs=classic#step-31-wix-installer) GUID must be the same.

I have updated the PR with the code.

yuyoyuppe commented:
@davidegiacometti we cannot reuse the runner's activator ID, since calling `RegisterAumidAndComServer` [will overwrite](https://github.com/App-vNext/Notifier/blob/a329d144a6e52bd4f094a4f5c0d034145577d0fe/src/AppVNext.Notifier.ConsoleUwp/DesktopNotificationManagerCompat.cs#L60-L69) the [runner's registry data](https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/installer/PowerToysSetup/Product.wxs#L294), meaning that all toast notification handling from the runner will be redirected to the PT Run service plugin executable, which isn't what we want.

If you take a look at how [notifications::register_application_id](https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/src/common/notifications.cpp#L124-L164) is implemented, all it's doing is creating a shortcut by writing 3 values to the registry. Perhaps we could just do the using dotnet's `
Registry.SetValue` API?

htcfreek commented:
> @davidegiacometti we cannot reuse the runner's activator ID, since calling `RegisterAumidAndComServer` [will overwrite](https://github.com/App-vNext/Notifier/blob/a329d144a6e52bd4f094a4f5c0d034145577d0fe/src/AppVNext.Notifier.ConsoleUwp/DesktopNotificationManagerCompat.cs#L60-L69) the [runner's registry data](https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/installer/PowerToysSetup/Product.wxs#L294), meaning that all toast notification handling from the runner will be redirected to the PT Run service plugin executable, which isn't what we want.
> 
> If you take a look at how [notifications::register_application_id](https://github.com/microsoft/PowerToys/blob/63ce7ca98167766821995a4da929e298a465ceb2/src/common/notifications.cpp#L124-L164) is implemented, all it's doing is creating a shortcut by writing 3 values to the registry. Perhaps we could just do the using dotnet's `
> Registry.SetValue` API?

The question is if we have to reload something to activate the newly created Registry keys? 🤔

yuyoyuppe commented:
@htcfreek no, we already use that functionality in bootstrapper.

TobiasSekan commented:
@davidegiacometti 

1. Can you show the service name behind the display name?
2. Can you show the startup-type in the second line too?

```
Widows Update (wuauserv)
Status: Stopped * Startup: Manual (Trigger Start)
```

davidegiacometti commented:
![image](https://user-images.githubusercontent.com/25966642/100009092-a930e980-2dce-11eb-8127-62592102eadf.png)

- @TobiasSekan I have added name and startup mode.
- Since PowerLauncher is not running elevated we are not able to creare registry key at runtime. At this point we need to register these keys during installation in the `Product.wxs`.
- In order to improve the UAC message is it fine to create a small C# console application as @htcfreek suggested? https://github.com/microsoft/PowerToys/pull/8076#issuecomment-730722000
We can pass service name and action (start/stop/restart) as CLI arguments.

davidegiacometti commented:
> * Since PowerLauncher is not running elevated we are not able to creare registry key at runtime. At this point we need to register these keys during installation in the `Product.wxs`.
> * In order to improve the UAC message is it fine to create a small C# console application as @htcfreek suggested? [#8076 (comment)](https://github.com/microsoft/PowerToys/pull/8076#issuecomment-730722000)
>   We can pass service name and action (start/stop/restart) as CLI arguments.

@crutkas when you get a sec can you give me a feedback on these?
First is about registering AUMID for notification.

davidegiacometti commented:
Ready for review! 🚀 
I have updated the gif in the first post.

- Change "windows command processor" UAC title would be a further improvement.
- Let me know if we want the services only under specific action key and if yes, which key to use.

Now there is a basic toast implementation (just show text at the moment) available to plugins through IPublicAPI.
Will update dev doc after merge for debugging notifications and required registry keys.

htcfreek commented:
> - Change "windows command processor" UAC title would be a further improvement.

Do we create a new issue/pr for that?

> - Let me know if we want the services only under specific action key and if yes, which key to use.

I think this would make sense because the service features is not as important as the program or file search. But I don't have an idea which key we should use.
I would suggest `svc`. But we only have one character for the action key.

davidegiacometti commented:
Will open a PR after the release.

We can use something like ! or _
Plugin manager would be a nice feature to handle this.

htcfreek commented:
Additional idea: Should we add a button to open `services.msc`?

davidegiacometti commented:
> Additional idea: Should we add a button to open `services.msc`?

I would like a context action for opening the services in `services.msc` but I think this can't be done.
But in **Task manager > Services tab > right click** there is just an **open services**, I think we can to the same.

davidegiacometti commented:
- Added `_` keyword for the plugin
- Added context action for open services
- Improved shortcuts
  - `Enter` start/stop service
  - `Ctrl+Shift+R` restart service
  - `Ctrl+Shift+O` open services

![image](https://user-images.githubusercontent.com/25966642/101989217-28fbf680-3c9f-11eb-861d-09a9ee8e91fa.png)

htcfreek commented:
Do wr get this reafy for 0.29 or do we need more time?

davidegiacometti commented:
If we want to have this PR merged for 0.29 and start to get feedback from users I am fine with it.
I have already tested installation and it should be fine.

htcfreek reviewed (approved):
The behaviour we have now looks good to me.
For the uac helper we can create a new issue and fix it with a new PR.

@crutkas 
Can we get this in for 0.29? What do you think?

crutkas commented:
I defer to @enricogior, i'm on vacation next week.

Jay-o-Way commented:
> I think a Windows 10 toast notification would be the best solution: it tells the user about the action, while light-dismissable. Other services in Windows use that as well (e.g. USB drive or Your Phone).
> 
> PowerToys Settings uses these toast notifications, might be re-usable?

Toasts should only be used when there is no GUI that can give a message - in whatever form that might be.

Jay-o-Way commented:
The end of the month is close. Would this PR make it into the next update?

htcfreek commented:
> The end of the month is close. Would this PR make it into the next update?

No. I think 0.29 is feature completed. Btw. you can look at the referenced project milestone (it's 0.31).

Jay-o-Way commented:
> > The end of the month is close. Would this PR make it into the next update?
> 
> No. I think 0.29 is feature completed. Btw. you can look at the referenced project milestone (it's 0.31).

True. It's just that many things have been pushed back. One or two days ago there were 44 issues "to-do" in 0.29, and now there are just 12. 🧙🏻‍♂️🎩

crutkas commented:
This needs WSX inclusion and YML for signing.  that can either be a must fix for 0.31 new issue or included here.

davidegiacometti commented:
> This needs WSX inclusion and YML for signing. that can either be a must fix for 0.31 new issue or included here.

Setup was already done. I have added Service Plugin DLLs to `pipeline.user.windows.yml`.

enricogior commented on .pipelines/pipeline.user.windows.yml:
```diff
@@ -120,6 +120,12 @@ build:
+            - 'modules\launcher\Plugins\Microsoft.Plugin.Service\PowerToysInterop.dll'
+            - 'modules\launcher\Plugins\Microsoft.Plugin.Service\Microsoft.PowerToys.Settings.UI.Lib.dll'
+            - 'modules\launcher\Plugins\Microsoft.Plugin.Service\ManagedTelemetry.dll'
+            - 'modules\launcher\Plugins\Microsoft.Plugin.Service\ManagedCommon.dll'
```
I understand they were already used before this PR, but `Microsoft.Plugin` is not a good prefix, it should be `Microsoft.PowerToys.Run.Plugin.`
And we should also stop using `Wox`.
We should consider fixing that.

davidegiacometti replied:
Related to https://github.com/microsoft/PowerToys/issues/2125 and https://github.com/microsoft/PowerToys/issues/2055
Should I use new names for this plugin and have different names in 0.31?

enricogior replied:
When adding new binaries, let's use the new naming convention, for the old one we will gradually update them in small PRs.

davidegiacometti replied:
@enricogior `Microsoft.PowerToys.Run.Plugin` or `Microsoft.PowerToys.Run.Plugins`?

enricogior replied:
`Microsoft.PowerToys.Run.Plugin`

enricogior commented on .pipelines/pipeline.user.windows.yml:
```diff
@@ -120,12 +120,7 @@ build:
-            - 'modules\launcher\Plugins\Microsoft.Plugin.Service\Microsoft.PowerToys.Settings.UI.Lib.dll'
-            - 'modules\launcher\Plugins\Microsoft.Plugin.Service\ManagedTelemetry.dll'
-            - 'modules\launcher\Plugins\Microsoft.Plugin.Service\ManagedCommon.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin\Microsoft.PowerToys.Run.Plugin.Service.dll'
```
I think that using the namespace for this folder is pointless, all the plugins in `modules\launcher\Plugins\` are `Microsoft.PowerToys.Run` plugins, we can just call it with the plugin name, in this case `Service`

enricogior replied:
We will update the folders names for the other plugins in a different PR.

davidegiacometti replied:
This is the pipeline that is working on output dir.
Should I remove the namespace from both output dir and install dir?

From
`\x64\Release\modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Service`
`C:\Program Files\PowerToys\modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Service`
to
`\x64\Release\modules\launcher\Plugins\Service`
`C:\Program Files\PowerToys\modules\launcher\Plugins\Service`

enricogior replied:
Yes that would be great and it would be used as a reference for fixing all the other plugins.

htcfreek replied:
@enricogior
Do we have a standard for the naming and ist this written down in the dev docs?

davidegiacometti commented:
This should be ready to review and merge.

htcfreek commented:
@davidegiacometti 
@crutkas suggested in #7951 to use `!` as action key for the service plugin and I like the idea. I think `!` is more intuitive for services as `_`. Thoughts?

enricogior commented:
@davidegiacometti 
I've updated the initial comment with the new checklist we are now using for PRs, as per @crutkas previous comment.

enricogior commented:
@davidegiacometti 
let's use `!` as action key.

davidegiacometti commented:
I have set `!` as action keyword.
Checklist on top is updated. The only missing things is update docs.
[End user](https://docs.microsoft.com/en-us/windows/powertoys/run): new shortcuts and action key in 0.31

enricogior commented:
@davidegiacometti 
there two conflicts to resolve, after that I'll do the final review and merge it. Thanks.

htcfreek commented:
@TobiasSekan
Do you open two new issues for the further improvements (open ToDo items)?

enricogior commented:
@davidegiacometti 
I tested it and it works as expected, nice work!
I found only one minor glitch that may not be cause by the plugin itself: after starting or stopping a service, reopening PT Run shows the last query with the previous status.
The service is running and I stop it:
![image](https://user-images.githubusercontent.com/3206696/103759119-802d8680-5013-11eb-8490-cdb9051edf47.png)

The service is stopped:
![image](https://user-images.githubusercontent.com/3206696/103758636-bddddf80-5012-11eb-8e43-622237e42644.png)

I reopen PT Run and it shows the previous status:
![image](https://user-images.githubusercontent.com/3206696/103758787-f67db900-5012-11eb-9bec-dc3405caade1.png)

I edit the query (deleting the last letter) and the status is updated:
![image](https://user-images.githubusercontent.com/3206696/103758864-17dea500-5013-11eb-8738-ca6cdcb01605.png)

Same thing when starting the service.

I don't think we should block this PR because of that, first because it might not be the plugin fault, secondly because it's a minor glitch that can be fixed in a new PR.

enricogior reviewed (approved):
LGTM!

davidegiacometti commented:
@htcfreek I am going to open an issue to track the improvement for the UAC popup.
The second TODO has been implemented as a generic "open services" button like task manager does.

@enricogior thank you and good catch! Should I open an issue to track this behaviour?

htcfreek commented:
> @htcfreek I am going to open an issue to track the improvement for the UAC popup.
> The second TODO has been implemented as a generic "open services" button like task manager does.

Please reference me and maybe this PR in the new issue. You can lable it as enhancement.

> @enricogior thank you and good catch! Should I open an issue to track this behaviour?

Imo this is a bug so I think we should open a new issue. I think it should be labeled as bug.

enricogior commented:
@davidegiacometti 
> Should I open an issue to track this behaviour?

Yes sure.

htcfreek commented:
@enricogior 
In your comment about the restart behaviour the buttons are missing on the third image (PT Run after reopen). Is this a bug or not captured on the image?

davidegiacometti commented:
> @enricogior
> In your comment about the restart behaviour the buttons are missing on the third image (PT Run after reopen). Is this a bug or not captured on the image?

It's a bug. Not sure if already tracked. Also in 0.29 sometimes if you have one result buttons disappear but I don't have a repro.

enricogior commented:
The icons appear as soon as you move the mouse over the result or you use the keyboard to select a result.
